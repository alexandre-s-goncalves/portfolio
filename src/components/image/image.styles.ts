import styled, {keyframes, css} from 'styled-components';

interface ExtraProps {
  $color?: string;
  $keepColors?: boolean;
  $height?: string;
  $width?: string;
  $cursor?: 'default' | 'pointer';
  $rotation?: number;
  $spinner?: boolean;
  $spinnerSpeed?: number;
}

const spin = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

const spinnerAnimation = (velocidade: number) => css`
  animation: ${spin} infinite ${velocidade}s linear;
`;

export const ContainerImage = styled.div<ExtraProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${({$width}) => $width};
  height: ${({$height}) => $height};
  box-sizing: border-box;
`;

export const ImageStyled = styled.img.attrs(
  (props: any) =>
    ({
      'data-spinner': props.$spinner ? 'true' : undefined,
    }) as any,
)<ExtraProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: ${({$cursor}) => ($cursor === 'default' ? 'none' : 'auto')};
  cursor: ${({$cursor}) => $cursor ?? 'default'};
  transform: rotate(${({$rotation = 0}) => $rotation}deg);
  ${({$spinner, $spinnerSpeed}) => $spinner && $spinnerSpeed && spinnerAnimation($spinnerSpeed)}
`;
