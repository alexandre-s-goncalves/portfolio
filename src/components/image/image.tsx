import React from 'react';
import {ContainerImage, ImageStyled} from './image.styles';

interface ImageProps {
  testId?: string;
  src: string;
  alt?: string;
  cursor?: 'default' | 'pointer';
  height?: number;
  width?: number;
  rotation?: number;
  spinner?: boolean;
  spinnerSpeed?: number;
  color?: string;
  keepColors?: boolean;
}

export const Image = ({
  testId,
  src,
  alt,
  cursor,
  height = 32,
  width = 32,
  rotation = 0,
  spinner = false,
  spinnerSpeed = 0,
  color,
  keepColors = false,
  ...props
}: Readonly<ImageProps>) => {
  const containerStyles: React.CSSProperties = {
    minWidth: width,
    minHeight: height,
    maxWidth: width,
    maxHeight: height,
  };

  return (
    <ContainerImage
      data-testid="container-image"
      $width={`${width}px`}
      $height={`${height}px`}
      style={containerStyles}
      {...props}>
      <ImageStyled
        data-testid={testId ? `${testId}-img` : 'img'}
        src={src}
        alt={alt}
        $cursor={cursor}
        $rotation={rotation}
        $spinner={spinner}
        $spinnerSpeed={spinnerSpeed}
        $color={color}
        $keepColors={keepColors}
      />
    </ContainerImage>
  );
};
