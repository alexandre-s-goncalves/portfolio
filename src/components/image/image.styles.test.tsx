import React from 'react';
import {render} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import * as S from './image.styles';

describe('image.styles', () => {
  test('ContainerImage should match the snapshot', () => {
    const {container} = render(<S.ContainerImage />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('ImageStyled default element and default styles', () => {
    const {container} = render(<S.ImageStyled />);
    const el = container.firstChild as HTMLElement;
    expect(el.nodeName).toBe('IMG');
    expect(el).toHaveStyle('cursor: default');
    expect(el).toHaveStyle('transform: rotate(0deg)');
  });

  test('ImageStyled should match snapshot', () => {
    const {container} = render(<S.ImageStyled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('ImageStyled renders an <img> for svg data src (keepColors=false)', () => {
    const svgData =
      "data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'><path d='M0 0h1v1z' style='fill:%23000;stroke:none'/></svg>";
    const {container, getByTestId} = render(
      <S.ImageStyled data-testid="i" src={svgData} $color="#ffffff" $rotation={0} />,
    );
    const el = getByTestId('i');
    expect(el.nodeName).toBe('IMG');
    expect((el as HTMLImageElement).src).toContain('data:image/svg+xml');
    expect(el).toHaveStyle('transform: rotate(0deg)');
  });

  test('ImageStyled renders an <img> when $keepColors=true (no DOM warning)', () => {
    const svgData =
      "data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'><path d='M0 0h1v1z' fill='%23000'/></svg>";
    const {getByTestId} = render(
      <S.ImageStyled data-testid="kc" src={svgData} $keepColors $color="#ffffff" $rotation={0} />,
    );
    const el = getByTestId('kc');
    expect(el.nodeName).toBe('IMG');
    expect((el as HTMLImageElement).src).toContain('data:image/svg+xml');
  });

  test('ImageStyled includes spinner animation when $spinner and $spinnerSpeed provided', () => {
    const svgData =
      "data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'><path d='M0 0h1v1z' /></svg>";
    const {getByTestId} = render(
      <S.ImageStyled
        data-testid="spinimg"
        src={svgData}
        $color="#ffffff"
        $spinner
        $spinnerSpeed={0.1}
        $rotation={0}
      />,
    );
    const img = getByTestId('spinimg');
    expect(img.getAttribute('data-spinner')).toBe('true');
  });
  test('ContainerImage applies provided width and height', () => {
    const {getByTestId} = render(<S.ContainerImage data-testid="c" $width="40px" $height="30px" />);
    const container = getByTestId('c');
    expect(container).toHaveStyle('width: 40px');
    expect(container).toHaveStyle('height: 30px');
  });

  test('ImageStyled default cursor and pointer-events when cursor not provided', () => {
    const {container} = render(<S.ImageStyled />);
    const img = container.firstChild as HTMLElement;
    expect(img).toHaveStyle('cursor: default');
  });

  test('ImageStyled respects cursor prop and rotation transform', () => {
    const {getByTestId} = render(
      <S.ImageStyled data-testid="i" $cursor="pointer" $rotation={45} />,
    );
    const img = getByTestId('i');
    expect(img).toHaveStyle('cursor: pointer');
    expect(img).toHaveStyle('transform: rotate(45deg)');
  });

  test('ImageStyled includes spinner animation when $spinner and $spinnerSpeed provided', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="spin" $spinner $spinnerSpeed={0.1} />);
    const img = getByTestId('spin');
    expect(img.getAttribute('data-spinner')).toBe('true');
  });

  test('ImageStyled pointer-events is "auto" when cursor is pointer', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="ptr" $cursor="pointer" />);
    const img = getByTestId('ptr');
    expect(img).toHaveStyle('pointer-events: auto');
    expect(img).toHaveStyle('cursor: pointer');
  });

  test('ImageStyled pointer-events is "none" when cursor is default', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="def" $cursor="default" />);
    const img = getByTestId('def');
    expect(img).toHaveStyle('pointer-events: none');
    expect(img).toHaveStyle('cursor: default');
  });

  test('ImageStyled does not include data-spinner attribute when $spinner is false', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="no-spin" $spinner={false} />);
    const img = getByTestId('no-spin');
    expect(img.getAttribute('data-spinner')).toBeNull();
  });

  test('ImageStyled with $color prop', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="colored" $color="#FF0000" />);
    const img = getByTestId('colored');
    expect(img).toBeInTheDocument();
  });

  test('ImageStyled with $keepColors prop', () => {
    const {getByTestId} = render(<S.ImageStyled data-testid="keep" $keepColors />);
    const img = getByTestId('keep');
    expect(img).toBeInTheDocument();
  });
});
