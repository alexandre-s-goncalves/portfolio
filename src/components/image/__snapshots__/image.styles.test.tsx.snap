// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image.styles ContainerImage should match the snapshot 1`] = `
.c0 {
  display: flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
}

<div
  class="c0"
/>
`;

exports[`image.styles ImageStyled should match snapshot 1`] = `
.c0 {
  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: auto;
  cursor: default;
  transform: rotate(0deg);
}

<img
  class="c0"
/>
`;
